plugin/pkg/replacer/replacer.go

```diff
@@ -41,168 +56,247 @@ var labels = []string{
-		return strconv.Itoa(state.Req.Opcode)
+		return strconv.AppendInt(b, int64(state.Req.Opcode), 10)
 	case headerReplacer + "do}":
-		return boolToString(state.Do())
+		return strconv.AppendBool(b, state.Do())
 	case headerReplacer + "bufsize}":
-		return strconv.Itoa(state.Size())
+		return strconv.AppendInt(b, int64(state.Size()), 10)
 	// Recorded replacements.
 	case "{rcode}":
 		if rr == nil {
-			return EmptyValue
+			return append(b, EmptyValue...)
 		}
-		rcode := dns.RcodeToString[rr.Rcode]
-		if rcode == "" {
-			rcode = strconv.Itoa(rr.Rcode)
+		if rcode := dns.RcodeToString[rr.Rcode]; rcode != "" {
+			return append(b, rcode...)
 		}
-		return rcode
+		return strconv.AppendInt(b, int64(rr.Rcode), 10)
 	case "{rsize}":
 		if rr == nil {
-			return EmptyValue
+			return append(b, EmptyValue...)
 		}
-		return strconv.Itoa(rr.Len)
+		return strconv.AppendInt(b, int64(rr.Len), 10)
 	case "{duration}":
 		if rr == nil {
-			return EmptyValue
+			return append(b, EmptyValue...)
 		}
-		return strconv.FormatFloat(time.Since(rr.Start).Seconds(), 'f', -1, 64) + "s"
+		secs := time.Since(rr.Start).Seconds()
+		return append(strconv.AppendFloat(b, secs, 'f', -1, 64), 's')
 	case headerReplacer + "rflags}":
 		if rr != nil && rr.Msg != nil {
-			return flagsToString(rr.Msg.MsgHdr)
+			return appendFlags(b, rr.Msg.MsgHdr)
 		}
-		return EmptyValue
-	}
-	return EmptyValue
-}
-
-// New makes a new replacer. This only needs to be called once in the setup and then call Replace for each incoming message.
-// A replacer is safe for concurrent use.
-func New() Replacer {
-	return Replacer{
-		valueFunc: value,
-		labels:    labels,
-	}
-}
-
-// Replace performs a replacement of values on s and returns the string with the replaced values.
-func (r Replacer) Replace(ctx context.Context, state request.Request, rr *dnstest.Recorder, s string) string {
-	for _, placeholder := range r.labels {
-		if strings.Contains(s, placeholder) {
-			s = strings.Replace(s, placeholder, r.valueFunc(state, rr, placeholder), -1)
-		}
-	}
-
-	// Metadata label replacements. Scan for {/ and search for next }, replace that metadata label with
-	// any meta data that is available.
-	b := strings.Builder{}
-	for strings.Contains(s, labelReplacer) {
-		idxStart := strings.Index(s, labelReplacer)
-		endOffset := idxStart + len(labelReplacer)
-		idxEnd := strings.Index(s[endOffset:], "}")
-		if idxEnd > -1 {
-			label := s[idxStart+2 : endOffset+idxEnd]
-
-			fm := metadata.ValueFunc(ctx, label)
-			replacement := EmptyValue
-			if fm != nil {
-				replacement = fm()
-			}
-
-			b.WriteString(s[:idxStart])
-			b.WriteString(replacement)
-			s = s[endOffset+idxEnd+1:]
-		} else {
-			break
-		}
-	}
-
-	b.WriteString(s)
-	return b.String()
-}
-
-func boolToString(b bool) string {
-	if b {
-		return "true"
+		return append(b, EmptyValue...)
+	default:
+		// TODO (CEV): consider panicking since this should be impossible
+		return append(b, EmptyValue...)
```

Comment:
can we return an error?

Reply:
This code will only be executed if there is a bug in the `parseFormat()` or `replacer.Replace()` logic and matches the previous behavior of the now removed `value()` function.  

Basically, this would be a hard programming error so I'm not sure there would be much value in propagating this to the user as there is nothing they can do about it (nor is there an easy way to propagate this error).

Personally, I would panic as I like to treat programming/logic errors harshly so that they do not make into production (I consider the behavior of unlocking an unlocked `sync.Mutex` to be a good example of this).  Additionally, if there is a bug here it is impossible to know if the logs being generated are correct.

**TLDR:** I don't think there is much value in returning an error - since this represents a logic bug.

Reply:
Comment removed.

Reply:
Note the comment has been removed, but is not showing up in the diff.